frontend/javascripts/dashboard/dataset/dataset_cache_provider.js

```diff
@@ -122,18 +123,20 @@ export default function DatasetCacheProvider({ children }: { children: Node }) {
       if (optionalUpdateIndex !== -1 || internalUpdateIndex !== -1) {
         const updatedDataset = await getDataset(datasetId);
 
-        if (optionalUpdateIndex !== -1 && datasetsToUpdate) {
-          const newDatasets = datasetsToUpdate.slice();
-          newDatasets[optionalUpdateIndex] = updatedDataset;
-          setDatasets(newDatasets);
-        }
-
         if (internalUpdateIndex !== -1) {
           const newDatasets = datasets.slice();
           newDatasets[internalUpdateIndex] = updatedDataset;
           datasetCache.set(newDatasets);
           if (!datasetsToUpdate) setDatasets(newDatasets);
         }
+
+        if (optionalUpdateIndex !== -1 && datasetsToUpdate) {
+          const newDatasets = datasetsToUpdate.slice();
+          const copyOfUpdatedDataset = (_.cloneDeep(updatedDataset): any);
+          copyOfUpdatedDataset.lastUsedByUser = newDatasets[optionalUpdateIndex].lastUsedByUser;
+          newDatasets[optionalUpdateIndex] = (copyOfUpdatedDataset: APIMaybeUnimportedDataset);
+          setDatasets(newDatasets);
+        }
```

Comment:
I think, the entire update logic can be simplified by using a `map` operation like this:

```js
if (datasetsToUpdate) {
  const newDatasets = datasetsToUpdate.map(dataset => {
    if (dataset.name === datasetId.name && ...) {
      const { lastUsedByUser } = dataset;
      return {
        ...updatedDataset,
        lastUsedByUser,
      }
    } else {
      return dataset;
    }
  })
}
```

That way, you don't have to remember indices and chop/join arrays :) The same goes for "internal dataset". You will probably need to always fetch the requested dataset (even if it does not appear in the datasets-arrays,) however, that shouldn't be a problem.

Reply:
Thx for the suggestion. The code looks way cleaner now, but for some reason flow does not like it. Would you mind taking another look?

Reply:
code looks great! for the record: I couldn't figure out how to tame flow. suppressing the error seems like an okay workaround for now :+1: